mm2src/mm2_bin_lib/src/mm2_native_lib.rs

```diff
@@ -69,52 +69,58 @@ pub unsafe extern "C" fn mm2_main(conf: *const c_char, log_cb: extern "C" fn(lin
 
     register_callback(FfiCallback::with_ffi_function(log_cb));
 
-    let (init_sender, init_receiver) = std::sync::mpsc::channel();
-
-    let rc = thread::Builder::new().name("lp_main".into()).spawn(move || {
-        let ctx_cb = &|ctx| CTX.store(ctx, Ordering::Relaxed);
+    let ctx_cb = &|ctx| CTX.store(ctx, Ordering::Relaxed);
+    let params = LpMainParams::with_conf(conf).log_filter(None);
+
+    let ctx = match catch_unwind(|| {
+        block_on(mm2_main::lp_main(
+            params,
+            &ctx_cb,
+            KDF_VERSION.into(),
+            KDF_DATETIME.into(),
+        ))
+    }) {
+        Ok(Ok(ctx)) => ctx,
+        Ok(Err(e)) => {
+            log!("MM2 initialization failed: {}", e);
+            LP_MAIN_RUNNING.store(false, Ordering::Relaxed);
+            return StartupResultCode::InitError as i8;
+        },
+        Err(err) => {
+            log!("MM2 initialization panicked: {:?}", any_to_str(&*err));
+            LP_MAIN_RUNNING.store(false, Ordering::Relaxed);
+            return StartupResultCode::InitError as i8;
+        },
+    };
 
-        let inner_sender = init_sender.clone();
+    // This allows us to use catch_unwind inside the lp_run thread despite MmCtx containing
+    // types with interior mutability (Mutex, RwLock, etc.) that aren't UnwindSafe.
+    // By passing just a numeric ID and recovering the actual context inside the
+    // catch_unwind block, we satisfy the compiler's safety requirements while properly
+    // handling potential panics.
+    let ctx_id = match ctx.ffi_handle() {
+        Ok(id) => id,
+        Err(e) => {
+            log!("MM2 thread setup failed: Failed to create FFI handle: {}", e);
+            LP_MAIN_RUNNING.store(false, Ordering::Relaxed);
+            return StartupResultCode::InitError as i8;
+        },
+    };
 
+    let rc = thread::Builder::new().name("lp_run".into()).spawn(move || {
         match catch_unwind(move || {
-            let params = LpMainParams::with_conf(conf).log_filter(None);
-
-            match block_on(mm2_main::lp_main(
-                params,
-                &ctx_cb,
-                KDF_VERSION.into(),
-                KDF_DATETIME.into(),
-            )) {
-                Ok(ctx) => {
-                    if let Err(err) = inner_sender.send(StartupResultCode::Ok) {
-                        log!(
-                            "Warning: Could not send initialization status - receiver disconnected error: {}",
-                            err
-                        );
-                    }
-
-                    block_on(mm2_main::lp_run(ctx));
+            let ctx = match MmArc::from_ffi_handle(ctx_id) {
+                Ok(ctx) => ctx,
+                Err(err) => {
+                    log!("Failed to recover context in thread: {:?}", err);
+                    return;
```

Comment:
nit: we could panic here instead to trigger an erroneous log from the catch_unwind instead of "MM2 thread completed normally"

Reply:
done